File: tests/bench/word_freq/word_freq.py
import os

_LEVEL = os.environ.get("BENCH_LEVEL", "hard")
SCALE = 16 if _LEVEL == "easy" else 4 if _LEVEL == "medium" else 1
WORDS = 200_000 // SCALE

text = ""
seed = 7
for _i in range(WORDS):
    seed = (seed * 1103515245 + 12345) % 2147483648
    text = text + f"w{seed % 500} "

counts = {}
start = 0
at = 0
n = len(text)
while at < n:
    if text[at] == " ":
        if at > start:
            word = text[start:at]
            counts[word] = counts.get(word, 0) + 1
        start = at + 1
    at += 1

total = 0
for _key, value in counts.items():
    total += value
print(len(counts))
print(total)

File: tests/bench/word_freq/WordFreq.java
// Same program as word_freq.jai.
import java.util.HashMap;
import java.util.Map;

public class WordFreq {
    // BENCH_LEVEL picks how much work this benchmark does: easy a sixteenth of it,
    // medium a quarter, hard (the default, and anything unrecognised) all of it.
    static long scale() {
        String l = System.getenv("BENCH_LEVEL");
        if ("easy".equals(l)) return 16;
        if ("medium".equals(l)) return 4;
        return 1;
    }

    static final long SCALE = scale();
    static final long WORDS = 200000L / SCALE;

    public static void main(String[] args) {
        StringBuilder builder = new StringBuilder();
        long seed = 7;
        for (long i = 0; i < WORDS; i++) {
            seed = (seed * 1103515245L + 12345L) % 2147483648L;
            builder.append("w").append(seed % 500L).append(" ");
        }
        String text = builder.toString();

        Map<String, Long> counts = new HashMap<>();
        int start = 0;
        int at = 0;
        int n = text.length();
        while (at < n) {
            if (text.charAt(at) == ' ') {
                if (at > start) {
                    String word = text.substring(start, at);
                    counts.put(word, counts.getOrDefault(word, 0L) + 1L);
                }
                start = at + 1;
            }
            at++;
        }

        long total = 0;
        for (Map.Entry<String, Long> entry : counts.entrySet()) {
            total += entry.getValue();
        }
        System.out.println(counts.size());
        System.out.println(total);
    }
}
